include/sentry.h

```diff
@@ -2058,6 +2075,94 @@ SENTRY_EXPERIMENTAL_API const char *sentry_sdk_name(void);
+ *
+ * See https://develop.sentry.dev/delightful-developer-metrics/
+ */
+struct sentry_metric_s;
+typedef struct sentry_metric_s sentry_metric_t;
+
+/**
+ * Creates a counter metric.
+ *
+ * Counters are one of the more basic types of metrics and can be
+ * used to count certain event occurrences.
+ *
+ * Counter value can only be incremented.
+ */
+SENTRY_EXPERIMENTAL_API sentry_metric_t *sentry_metrics_new_increment(
+    const char *key, double value);
+SENTRY_EXPERIMENTAL_API sentry_metric_t *sentry_metrics_new_increment_n(
+    const char *key, size_t key_len, double value);
+
+/**
+ * Creates a distribution metric.
+ *
+ * Distributions track a list of values over time on which
+ * aggregations like `p90`, `max`, `min`, `avg` can be performed.
+ */
+SENTRY_EXPERIMENTAL_API sentry_metric_t *sentry_metrics_new_distribution(
+    const char *key, double value);
+SENTRY_EXPERIMENTAL_API sentry_metric_t *sentry_metrics_new_distribution_n(
+    const char *key, size_t key_len, double value);
+
+/**
+ * Creates a gauge metric.
+ *
+ * Gauges track a value that can go up and down on which
+ * aggregations like `max`, `min`, `avg`, `sum` and `count` can be performed.
+ *
+ * Gauges can be represented in a more space-efficient way than distributions,
+ * but they can't be used to get percentiles.
+ */
+SENTRY_EXPERIMENTAL_API sentry_metric_t *sentry_metrics_new_gauge(
+    const char *key, double value);
+SENTRY_EXPERIMENTAL_API sentry_metric_t *sentry_metrics_new_gauge_n(
+    const char *key, size_t key_len, double value);
+
+/**
+ * Creates a set metric.
+ *
+ * Sets are useful for looking at unique occurrences and counting
+ * the unique elements that were added.
+ *
+ * Sets track a set of values on which aggregations such as `count_unique`
+ * can be performed.
+ */
+SENTRY_EXPERIMENTAL_API sentry_metric_t *sentry_metrics_new_set(
+    const char *key, int32_t value);
+SENTRY_EXPERIMENTAL_API sentry_metric_t *sentry_metrics_new_set_n(
+    const char *key, size_t key_len, int32_t value);
+
+/**
+ * Sets a tag on a metric to the given string value.
+ *
+ * Tags longer than 200 bytes will be truncated.
+ *
+ * The metric should not be mutated by other functions while a tag is being set
+ * on it.
+ */
+SENTRY_EXPERIMENTAL_API void sentry_metric_set_tag(
+    sentry_metric_t *metric, const char *tag, const char *value);
+SENTRY_EXPERIMENTAL_API void sentry_metric_set_tag_n(sentry_metric_t *metric,
+    const char *tag, size_t tag_len, const char *value, size_t value_len);
+
+/**
+ * Sets a measurement unit of the given metric.
+ *
+ * See https://develop.sentry.dev/sdk/metrics/#units.
+ */
+SENTRY_EXPERIMENTAL_API void sentry_metric_set_unit(
+    sentry_metric_t *metric, const char *unit);
+SENTRY_EXPERIMENTAL_API void sentry_metric_set_unit_n(
+    sentry_metric_t *metric, const char *unit, size_t unit_len);
+
+/**
+ * Captures a metric to be sent to Sentry.
+ */
+SENTRY_EXPERIMENTAL_API void sentry_metrics_capture(sentry_metric_t *metric);
```

Comment:
As mentioned in our call, we should make this cheaper for the user. The way it is currently implemented, for C/C++ users not familiar with the other SDKs, it probably seems weird to create an opaque handle for something that requires creating it with every change of the metric.

We should reuse the `sentry_metric_t` handle passed to the user or - preferably - hide the complexity behind another interface and replace `sentry_metric_t` with "normal" values. For the latter, I suggest providing a single function for each metric (e.g., `sentry_metric_new_counter()`) that takes tags as a variadic parameter, like so:

```C
sentry_metrics_new_counter(char *key, double value, char *unit, ...)
{
    size_t key_len = key ? strlen(key) : 0;
    sentry_value_t metric
        = sentry__value_metric_new_n((sentry_slice_t) { key, key_len });
    if (sentry_value_is_null(metric)) {
        return;
    }
    sentry_value_set_by_key(
        metric, "type", sentry__metrics_type_to_string(SENTRY_METRIC_COUNTER));
    sentry_value_set_by_key(metric, "value", sentry_value_new_double(value));

    if (unit) {
        sentry_value_set_by_key(
            metric, "unit", sentry_value_new_string_n(unit, strlen(unit)));
    }

    va_list args;
    va_start(args, unit);

    const char *tag;
    const char *tag_value;
    while (1) {
        tag = va_arg(args, const char *);
        if (tag == NULL) {
            break;
        }
        tag_value = va_arg(args, const char *);
        set_tag(metric, tag, tag_value);
    }
    va_end(args);

    SENTRY_WITH_METRICS_AGGREGATOR(aggregator)
    {
        sentry__metrics_aggregator_add_value(aggregator, metric);
    }
}
```

As you can see, there is no need to use the `sentry_metric_t` struct because we can use Native SDK values directly. In this case, we also eliminate the issue of keeping the value alive until it is sent because it will automatically be deallocated in the transport. The changes to `sentry_value_t` spread to `sentry__metrics_aggregator_add` and `sentry__metrics_get_metric_bucket_key`, but not much further because most functions in the metrics module work with Native SDK values.